source/vtk/utilities.cc

```diff
@@ -0,0 +1,235 @@
+// ------------------------------------------------------------------------
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2025 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// Part of the source code is dual licensed under Apache-2.0 WITH
+// LLVM-exception OR LGPL-2.1-or-later. Detailed license information
+// governing the source code and code contributions can be found in
+// LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
+//
+// ------------------------------------------------------------------------
+#include <deal.II/vtk/utilities.h>
+
+#include <string>
+#include <vector>
+
+#ifdef DEAL_II_WITH_VTK
+#  include <deal.II/distributed/fully_distributed_tria.h>
+
+#  include <deal.II/dofs/dof_renumbering.h>
+
+#  include <deal.II/fe/fe.h>
+#  include <deal.II/fe/fe_dgq.h>
+#  include <deal.II/fe/fe_nothing.h>
+#  include <deal.II/fe/fe_q.h>
+#  include <deal.II/fe/fe_system.h>
+
+#  include <deal.II/grid/grid_in.h>
+#  include <deal.II/grid/grid_out.h>
+#  include <deal.II/grid/grid_tools.h>
+#  include <deal.II/grid/tria_accessor.h>
+#  include <deal.II/grid/tria_description.h>
+#  include <deal.II/grid/tria_iterator.h>
+
+#  include <vtkCellData.h>
+#  include <vtkCleanUnstructuredGrid.h>
+#  include <vtkDataArray.h>
+#  include <vtkPointData.h>
+#  include <vtkSmartPointer.h>
+#  include <vtkUnstructuredGrid.h>
+#  include <vtkUnstructuredGridReader.h>
+
+#  include <stdexcept>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace VTKWrappers
+{
+  template <int dim, int spacedim>
+  void
+  read_vtk(const std::string            &vtk_filename,
+           Triangulation<dim, spacedim> &tria,
+           const bool                    cleanup)
+  {
+    auto reader = vtkSmartPointer<vtkUnstructuredGridReader>::New();
+    {
+      // check that the file exists
+      std::ifstream file(vtk_filename);
+      AssertThrow(file.good(),
+                  ExcMessage("VTK file not found: " + vtk_filename));
+    }
+    reader->SetFileName(vtk_filename.c_str());
+    reader->Update();
+    vtkUnstructuredGrid *grid = reader->GetOutput();
+    AssertThrow(grid, ExcMessage("Failed to read VTK file: " + vtk_filename));
+
+    auto cleaner = vtkSmartPointer<vtkCleanUnstructuredGrid>::New();
+
+    // Cleanup the triangulation if requested
+    if (cleanup)
+      {
+        cleaner->SetInputData(grid);
+        cleaner->Update();
+        grid = cleaner->GetOutput();
+        AssertThrow(grid,
+                    ExcMessage("Failed to clean VTK file: " + vtk_filename));
```

Comment:
Can these cleanup functions handle, e.g., merging vertices within a specific tolerance?

Reply:
Yes. I'll add the tolerance (this afternoon).